### PR TITLE
Modification of ReaxFF

### PR DESCRIPTION
**Pull request by mkanski**
## Purpose

Recently we have published a paper [1] describing a modified version of ReaxFF for hydrocarbons which is faster than the standard parametrization, while maintaining similar agreement with reference data. The speed-up was achieved by excluding charge and reducing cutoff for non-bonded interactions. Our parametrization works with the current ReaxFF implementation in LAMMPS, but in order to achieve full speed-up a simple modification of the source code is needed. 
If it is possible, I would like to distribute our parametrization with LAMMPS. 

The C/H/O version should be available later this year.

EDIT:

I've added an example and the tabulated correction for close-range interactions.

## Author(s)

Michal Kanski, Jagiellonian University

## Backward Compatibility

N/A

## Implementation Notes

This pull request introduces a small modification of how the cutoff for ReaxFF is chosen. New version takes into account a bonded cutoff instead of the doubled bonded cutoff. 

This change does not change the cutoff for other parametrizations, because their non-bonded cutoff is always equal to 10A. 
In order to check that, a simple simulation of polyethylene crystal [2,3] was performed. The results for ffield.reax.cho are the same for standard and modified LAMMPS [4,5].

## Post Submission Checklist

_Please check the fields below as they are completed_
- [x] The feature or features in this pull request is complete
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [x] One or more example input decks are included
- [ ] The source code follows the LAMMPS formatting guidelines

## Further Information, Files, and Links

[1] DOI: 10.1021/acs.jpclett.7b03155
[2] [input.in.txt](https://github.com/lammps/lammps/files/1611871/input.in.txt)
[3] [PE_crystal.dat.txt](https://github.com/lammps/lammps/files/1611872/PE_crystal.dat.txt)
[4] [log.lammps.txt](https://github.com/lammps/lammps/files/1612208/log.lammps.txt)
[5] [log.lammps_modified.txt](https://github.com/lammps/lammps/files/1612207/log.lammps_modified.txt)

### CODE REVIEW

File: src/USER-REAXC/pair_reaxc.cpp [hunk 1]
```diff
@@ -387,7 +387,7 @@ void PairReaxC::init_style( )
   neighbor->requests[irequest]->newton = 2;
   neighbor->requests[irequest]->ghost = 1;
 
-  cutmax = MAX3(control->nonb_cut, control->hbond_cut, 2*control->bond_cut);
+  cutmax = MAX3(control->nonb_cut, control->hbond_cut, control->bond_cut);
```
Comment (akohlmey) [inline src/USER-REAXC/pair_reaxc.cpp]: Please note, that the same change would need to be made in USER-OMP/pair_reaxc_omp.cpp as it overrides init_style().

**Comment by akohlmey** [comment]
How does this change relate to the optimization in commit c3aa705d0409ddf9457cd1ba86e43ef10fc42d6e?

**Comment by mkanski** [comment]
If I understand correctly, this change and the one from commit c3aa705 optimize different things.
The change in the commit optimizes neighbours for both local and ghost atoms. The change I propose, reduces the number of ghost atoms for ReaxFF potentials with non-bonded cutoff lower than 10A (or more generally, lower than doubled cutoff for bonded interactions), thus reducing amount of calculations.

For the test simulation (input files are in the first post) the speed-up is about 15%:

Log with standard LAMMPS: [log.ci-ReaxFF.txt](https://github.com/lammps/lammps/files/1619388/log.ci-ReaxFF.txt)
Log with the change: [log.ci-ReaxFF_modified.txt](https://github.com/lammps/lammps/files/1619389/log.ci-ReaxFF_modified.txt)

File: src/USER-REAXC/pair_reaxc.cpp (hunk 1 above)
Comment (mkanski) [reply src/USER-REAXC/pair_reaxc.cpp]: If the pull request is accepted, I can make the change in the USER-OMP version of ReaxFF.

Comment (akohlmey) [reply src/USER-REAXC/pair_reaxc.cpp]: You have it backwards. We won't merge a pull request with known inconsistencies.

Comment (mkanski) [reply src/USER-REAXC/pair_reaxc.cpp]: Of course, I've made the change.

**Comment by stanmoore1** [comment]
@mkanski I haven't had a chance to look at this in detail yet, but do you know why the factor of 2 was originally included in the code, i.e. is this change going to break other systems?

**Comment by akohlmey** [comment]
@stanmoore1 @mkanski 
what i would be most concerned about when reducing ghost and communication cutoff, is handling of angle and particularly dihedral interactions. i would assume a good test case could be a strained polyethylene chain (make a carbon on one side immobile and full on the other side with fix addforce).

**Comment by mkanski** [comment]
@stanmoore1 
In the original paper by Aktulga et al. [1] they used factor of 3, in order to ensure that all atoms in a dihedral angle are in a neighbour list. I assume the doubled bonded cutoff in LAMMPS is an optimization of that. 
The change I propose will make a difference only for ReaxFF parametrizations where non-bonded cutoff is lower than doubled bonded cutoff. In almost all parameterizations these cutoffs are equal 10A and 5A for non-bonded and bonded interactions, respectively, so this modification won't change anything.

@akohlmey 
I did the simulations you had proposed [2]. The forces start diverging at 1370th timestep . However, if I restart the simulations from timestep where the difference between forces becomes significant (for example 2859th timestep) [3], the differences at the beginning are in order of 10^-13 kcal/mol/A [4].

I did also additional simulations [5] where a polyethylene chain is simply elongated by fix move. The results are identical. Therefore I think that differences in [2] are due to round-off errors.


[1] Aktulga, Fogarty, Pandit, Grama, Parallel Computing, 38, 245-259 (2012).
[2] [polyethylene_chain_addforce_input_files.zip](https://github.com/lammps/lammps/files/1623073/polyethylene_chain_addforce_input_files.zip)
[3] [polyethylene_chain_addforce_restart_input_files.zip](https://github.com/lammps/lammps/files/1623067/polyethylene_chain_addforce_restart_input_files.zip)
[4] 
force_error


[5] [polyethylene_chain_fix_move.zip](https://github.com/lammps/lammps/files/1622786/polyethylene_chain_fix_move.zip)

**Comment by stanmoore1** [comment]
I ran the `src/USER-REAXC/pair_reaxc.cpp` cutoff change through our regression tests. On 4 processors, there was no difference in thermo output between using the different cutoffs.

**Comment by akohlmey** [comment]
@stanmoore1 @mkanski after seeing so many successful tests, i am inclined to recommend this change for inclusion. But i would feel better, if there was a warning added. I just added a proposal for such a warning to the pull request. Please have a look and improve or modify as needed.

**Comment by mkanski** [comment]
Thank you for your help. 
@akohlmey The warning is reasonable. Thanks again.

**Pull request merged**